src/serialize.h

```diff
@@ -573,6 +573,64 @@ template<typename I> class BigEndianWrapper
 //! Automatically construct a BigEndianWrapper around the argument.
 template<typename I> static inline BigEndianWrapper<I> BigEndian(I& i) { return BigEndianWrapper<I>(i); }
 
+/** Serialization wrapper for custom-element vectors.
+ *
+ * This allows (de)serialization of vectors of type V while using a custom
+ * serializer W for the entries inside.
+ *
+ * For const-correctness, the W parameter is a template itself.
+ * It is instantiated as W<const V::value_type> when serializing, and as
+ * W<V::value_type> when deserializing.
+ *
+ * Example:
+ *   struct X {
+ *     std::vector<uint64_t> v;
+ *     SERIALIZE_METHODS(obj) { READWRITE(VectorApply<VarIntWrapper>(obj.v)); }
+ *   };
+ * will define a struct that contains a vector of uint64_t, which is serialized
+ * as a vector of VarInt-encoded integers.
+ *
+ * V is not required to be an std::vector type. It works for any class that
+ * exposes a value_type, iteration, and resize method that behave like vectors.
+ */
+template<template <typename> class W, typename V> class VectorApplyWrapper
+{
+protected:
+    typedef typename V::value_type value_type;
+    V& m_vector;
+public:
+    explicit VectorApplyWrapper(V& vector) : m_vector(vector) {}
+
+    template<typename Stream>
+    void Serialize(Stream& s) const
+    {
+        WriteCompactSize(s, m_vector.size());
+        for (const value_type& x : m_vector) {
+            s << W<const value_type>(x);
+        }
+    }
+
+    template<typename Stream>
+    void Unserialize(Stream& s)
+    {
+        m_vector.clear();
+        unsigned int nSize = ReadCompactSize(s);
+        unsigned int nMid = 0;
+        while (nMid < nSize) {
+            nMid += 5000000 / sizeof(value_type);
+            if (nMid > nSize) {
+                nMid = nSize;
+            }
+            m_vector.resize(nMid);
+            for (value_type& x : m_vector) {
+                s >> W<value_type>(x);
```

Comment:
In commit "Add custom vector-element serialization wrapper"

Doesn't this overwrite elements in the front of the vector each time through the while loop?

Reply:
Yes, this was bogus. Thanks for pointing that out; fixed by rewriting,